### ISSUE TITLE
TODO: way to keep scripts updated

### ISSUE THREAD

**Issue by wiiaboo** [jb-alvarado/media-autobuild_suite#162]
- Use `wget -n` to get files if they have been updated
 - easy to code (just grab list of files and check if they exist and if server has newer version)
 - depends on file modification timestamps, not on file hashes
 - fails if file doesn't exist on server
 - fails if local file was edited more recently than on server

- Use git
  - bit more difficult to code
  - uses file hashes, so doesn't depend on timestamps
  - local changes can be kept by making a commit and then rebasing or by `git stash`ing them or exporting them to a .diff and then the user decides if he wants them applied or not
  - won't work if server is down but can be used to `git checkout` to the last commit if the files have been deleted
  - if the suite already was a clone, it would just check for updates and run as usual

**Comment by jb-alvarado** [comment]
I have no try it, but maybe it works also to pipe wget to diff ?

**Comment by wiiaboo** [comment]
>I have no try it, but maybe it works also to pipe wget to diff ?

At that point it would work like git way, no?

We could assume first build is always updated and user got the whole archive through zip.
That way we could do the git stuff in update.sh.

We could also do both. Use wget if the files don't exist and then use git for future updates or if user deletes `build`.

**Comment by jb-alvarado** [comment]
Yes that would be almost the same then git. But you have write the git code is a bit more difficult. So I thought just wget/diff is maybe more easy. Using a combination of wget end git sounds good to me, but the problem what I have a bit with git, is that you must have a git checkout from the repo. Not only the zip file.

**Comment by wiiaboo** [comment]
You just have to `git clone` to `/build` and then copy the `.git` folder to `/trunk` and it's turned into a git repo. Very easy.

**Comment by jb-alvarado** [comment]
Hehe, ok, yes that is really easy.

**Comment by wiiaboo** [comment]
What's hard is writing all that in the .bat, so that's why you could get missing scripts using wget in the .bat and the git part in update.sh

**Comment by jb-alvarado** [comment]
Yes that sounds the best way for me.

**Comment by wiiaboo** [comment]
I've changed the name of the branch to "update" and corrected your commit to not remove `--enable-libtesseract`.
Also added a WIP git update code.

**Comment by jb-alvarado** [comment]
That is good - thank you! I It was a checkout before you add tesseract. I also thought about your idea to checkout the repo in the build dir and move the .git to trunk. In that way it always will pull the readme, license files and the doc folder. Or is there a way to say to git, only update certain files?

**Comment by wiiaboo** [comment]
You can `git checkout` specific files. The readme and doc are important, though. It's good that people get them.

**Comment by jb-alvarado** [comment]
That is good, and of course the readme and doc can stay there.

**Comment by wiiaboo** [comment]
Btw, in light of FFmpeg's aac encoder soon being considered not experimental, I'm thinking of removing some options from the default, where FFmpeg's native encoder/decoder are considered high quality/lossless and spec-compliant, specifically libopenjpeg and libutvideo.

**Comment by jb-alvarado** [comment]
I think that is ok, when the internal have the same quality then the external libs. You also talk about fdk-aac?

**Comment by wiiaboo** [comment]
There haven't been listening tests to confirm that yet. I tried it and it's close in quality to fdk-aac but still not there. Certainly higher than faac, though.

**Comment by jb-alvarado** [comment]
Nice!

**Comment by wiiaboo** [comment]
utvideo-wise, ffmpeg does lossless tests every time they run a build on their servers, so any difference between native and libutvideo would be speed (native is faster than libutvideo) or lack of features.
openjpeg-wise, I think libopenjpeg decoder is considered experimental when compiled from openjpeg2.

**Comment by wiiaboo** [comment]
Btw, had another idea now.
What do you think of separating msys2 decision from the others and installing the base system right away after that choice and leave the others to be chosen through a bash script that reads what options do exist so people don't have to repeat choices they don't need?

**Comment by wiiaboo** [comment]
This make work with batch too, after all: https://github.com/jb-alvarado/media-autobuild_suite/commit/b3613793d2f5b0d8712ce1ae3bab4509fcf02357
Haven't tested yet.

**Comment by jb-alvarado** [comment]
I general I would prefer to choice everything at the begin, so you can start the script and then let the computer running. And when you come back everything is done. On the other hand I like your idea to, that it is more easy to add new choices later, it makes it much more flexible. Maybe we can run bash truth batch?

**Comment by wiiaboo** [comment]
I tried today to do it with batch but was a bit sketchy. I don't think it would be much better with bash either.
I'm trying a build with deleted /locals and it seems to work fine other than a problem with having an older mintty without `-d`. Other than that, I think it's safe to rebase and merge everything until now.
I don't think there is a way to detect it automatically other than asking users to run mintty and `pacman -Sy mintty` before running the script.

**Comment by jb-alvarado** [comment]
Yeah, the question is what will be happen, when old mintty start with missing `-d` option. Maybe not much, and it still run :). You can merge it, and the we see what people post.

**Comment by wiiaboo** [comment]
It doesn't run, actually. But there's no way to detect it, at least with errorlevel.
Only way would be:
 - user manually running `pacman -Sy mintty`
 - user deleting /msys folder
 - running all scripts except compile.sh directly with bash, losing logging.

**Comment by jb-alvarado** [comment]
I thought for a moment to check with pacman the version (when is possible), but would be only a temporary option, and this I don't like so much. The is better to say people to run the pacman update.

**Comment by wiiaboo** [comment]
OK, finished now. My build still works and full build works too, should be safe.

**Comment by jb-alvarado** [comment]
Yes I can confirm that it works.

**Comment by wiiaboo** [comment]
The update feature?

**Comment by jb-alvarado** [comment]
I have override the scripts and run it. I had to update mintty, and after that it have work. But now I see that it can not remove libopus, because of opusfile. I have not a git checkout, only the scripts, so I am not sure that it makes also a update with the scripts it self.

**Comment by wiiaboo** [comment]
I tried it from just the .bat without anything else and it more or less worked.
The .bat closed after the wget for some reason, didn't reproduce after running from cmd.exe.
The changes weren't applied because it was a fresh checkout, so I have to change that. And the diff had useless stuff because there weren't any files besides the .bat, I have to filter the diff to just modifications(?)
opus-tools needed flac and I turned it off so there's a fix coming for that too.
